examples/all-clusters-app/esp32/main/Rpc.cpp

```diff
@@ -0,0 +1,328 @@
+        scan_config.scan_type   = static_cast<wifi_scan_type_t>(request.active_scan);
+        if (request.active_scan)
+        {
+            scan_config.scan_time.active.min = request.scan_time_min_ms;
+            scan_config.scan_time.active.max = request.scan_time_max_ms;
+        }
+        else
+        {
+            scan_config.scan_time.passive = request.scan_time_min_ms;
+        }
+
+        auto err = esp_wifi_scan_start(&scan_config, true /* block */);
+        if (ESP_OK != err)
+        {
+            ESP_LOGI(TAG, "Error starting scan: %d", err);
+            return;
+        }
+
+        // Output scan results
+        uint16_t num_scan_records = kScanRecordsMax;
+        err                       = esp_wifi_scan_get_ap_records(&num_scan_records, scan_records);
+        if (ESP_OK != err)
+        {
+            ESP_LOGI(TAG, "Error getting scanned APs: %d", err);
+            num_scan_records = 0;
+        }
+        ESP_LOGI(TAG, "%d", num_scan_records);
+        out_scan_records.aps_count = num_scan_records;
+
+        for (size_t i = 0; i < num_scan_records; ++i)
+        {
+            out_scan_records.aps[i].ssid.size = std::min(sizeof(scan_records[i].ssid), sizeof(out_scan_records.aps[i].ssid.bytes));
+            memcpy(out_scan_records.aps[i].ssid.bytes, scan_records[i].ssid, out_scan_records.aps[i].ssid.size);
+            out_scan_records.aps[i].bssid.size =
+                std::min(sizeof(scan_records[i].bssid), sizeof(out_scan_records.aps[i].bssid.bytes));
+            memcpy(out_scan_records.aps[i].bssid.bytes, scan_records[i].bssid, out_scan_records.aps[i].bssid.size);
+            out_scan_records.aps[i].security_type = static_cast<chip_rpc_WIFI_SECURITY_TYPE>(scan_records[i].authmode);
+            out_scan_records.aps[i].channel       = scan_records[i].primary;
+            auto found_channel                    = kChannelToFreqMap.find(scan_records[i].primary);
+            out_scan_records.aps[i].frequency     = (found_channel ? found_channel->second : 0);
+            out_scan_records.aps[i].signal        = scan_records[i].rssi;
+        }
+        writer.Write(out_scan_records);
+        writer.Finish();
+    }
+
+    pw::Status StopScan(ServerContext &, const pw_protobuf_Empty & request, pw_protobuf_Empty & response)
+    {
+        esp_wifi_scan_stop();
+        return pw::OkStatus();
+    }
+
+    pw::Status Connect(ServerContext &, const chip_rpc_ConnectionData & request, chip_rpc_ConnectionResult & response)
+    {
+        char ssid[sizeof(wifi_config_t().sta.ssid)];
+        char password[sizeof(wifi_config_t().sta.password)];
+        memcpy(ssid, request.ssid.bytes, std::min(sizeof(ssid), static_cast<size_t>(request.ssid.size)));
+        memcpy(password, request.secret.bytes, std::min(sizeof(password), static_cast<size_t>(request.secret.size)));
+        return SetWiFiStationProvisioning(ssid, password) == CHIP_NO_ERROR ? pw::OkStatus() : pw::Status::Unknown();
```

Comment:
`SetWiFiStationProvisioning` expects null-terminated input.  I don't see why these things would be null-terminated....

Reply:
I am literally just looking at "memcpy" calls and finding bugs every time I do.  :(

Reply:
Thanks again... Clearly I rushed this more then I should have. Appreciate the review.